airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/state/MemoryManager.kt

```diff
@@ -17,31 +19,49 @@ import kotlin.concurrent.withLock
  * TODO: Some degree of logging/monitoring around how accurate we're actually being?
  */
 @Singleton
-class MemoryManager {
-    private val availableMemoryBytes: Long = Runtime.getRuntime().maxMemory()
+class MemoryManager(availableMemoryProvider: AvailableMemoryProvider) {
+    private val totalMemoryBytes: Long = availableMemoryProvider.availableMemoryBytes
     private var usedMemoryBytes = AtomicLong(0L)
-    private val memoryLock = ReentrantLock()
-    private val memoryLockCondition = memoryLock.newCondition()
+    private val mutex = Mutex()
+    private val syncChannel = Channel<Unit>(Channel.UNLIMITED)
 
+    val remainingMemoryBytes: Long
+        get() = totalMemoryBytes - usedMemoryBytes.get()
+
+    /* Attempt to reserve memory. If enough memory is not available, waits until it is, then reserves. */
     suspend fun reserveBlocking(memoryBytes: Long) {
-        memoryLock.withLock {
-            while (usedMemoryBytes.get() + memoryBytes > availableMemoryBytes) {
-                memoryLockCondition.await()
+        if (memoryBytes > totalMemoryBytes) {
+            throw IllegalArgumentException(
+                "Requested ${memoryBytes}b memory exceeds ${totalMemoryBytes}b total"
+            )
+        }
+
+        mutex.withLock {
+            while (usedMemoryBytes.get() + memoryBytes > totalMemoryBytes) {
```

Comment:
yeah, I'm pretty sure this is correct. 
Just as an interesting case, we could also do without the AtomicLong and a volatile would suffice :)
Thank you for fixing it.